consensus-types/types/validator.go

```diff
@@ -317,3 +318,52 @@ func (v Validator) GetWithdrawableEpoch() math.Epoch {
 func (v Validator) GetWithdrawalCredentials() WithdrawalCredentials {
 	return v.WithdrawalCredentials
 }
+
+// Status returns the current validator status based on its set epoch values.
+// This function taken from Prysm:
+// https://github.com/prysmaticlabs/prysm/blob/0229a2055e6349655a471b2427f349e40c275cee/beacon-chain/rpc/eth/helpers/validator_status.go#L31
+func (v *Validator) Status(currentEpoch math.Epoch) (string, error) {
+	activationEpoch := v.GetActivationEpoch()
+	activationEligibilityEpoch := v.GetActivationEligibilityEpoch()
+	farFutureEpoch := math.Epoch(constants.FarFutureEpoch)
+	exitEpoch := v.GetExitEpoch()
+	withdrawableEpoch := v.GetWithdrawableEpoch()
+
+	// Status: pending
+	if activationEpoch > currentEpoch {
+		if activationEligibilityEpoch == farFutureEpoch {
+			return constants.ValidatorStatusPendingInitialized, nil
+		} else if activationEligibilityEpoch < farFutureEpoch {
+			return constants.ValidatorStatusPendingQueued, nil
+		}
+	}
+
+	// Status: active
+	if activationEpoch <= currentEpoch && currentEpoch < exitEpoch {
+		if exitEpoch == farFutureEpoch {
+			return constants.ValidatorStatusActiveOngoing, nil
+		} else if exitEpoch < farFutureEpoch {
+			if v.IsSlashed() {
+				return constants.ValidatorStatusActiveSlashed, nil
+			}
+			return constants.ValidatorStatusActiveExiting, nil
+		}
+	}
+
+	// Status: exited
+	if exitEpoch <= currentEpoch && currentEpoch < withdrawableEpoch {
+		if v.IsSlashed() {
+			return constants.ValidatorStatusExitedSlashed, nil
+		}
+		return constants.ValidatorStatusExitedUnslashed, nil
+	}
+
+	// Status: withdrawal
+	if withdrawableEpoch <= currentEpoch {
+		if v.GetEffectiveBalance() != math.Gwei(0) {
+			return constants.ValidatorStatusWithdrawalPossible, nil
+		}
+		return constants.ValidatorStatusWithdrawalDone, nil
+	}
+	return "", errors.New("invalid validator status")
```

Comment:
@berachain/core not sure I am happy with this choice here.
This case should only happen if the beacon state is messed up and not if above is possible (e.g. for a validator exitHeight < activationHeight).
However if this ever happens, this error would stop querying any validator, because of the error.
I know prysm does the same but should we instead log an error and let the other validators be returned?

Reply:
I actually thinks returning error here is good. If we see this error returned from the API, it’s a strong sign that the beacon state registry of validators needs a fixing

Reply:
I agree, but we would not see which validator is broken right? The API will return an error and no clear indications of where is the error.
Maybe we should at least decorate the final error with some info about the faulty validators

Reply:
Yeah can patch this error message with validator index and/or pubkey

Reply:
Introduced typed error [here](https://github.com/berachain/beacon-kit/pull/2467/commits/eb07e3f488ff80ecdcead91cccfd0f70f54a99b2)